### PR TITLE
DO NOT MERGE Create Scala.js-defined JS class values lazily.

### PR DESCRIPTION
**Pull request by sjrd**
This will be necessary to implement features which require user-defined code to be executed at class-definition-time. Since such code can have arbitrary cyclic dependencies on anything else, it is absolutely necessary that everything else is defined or lazily loaded.

The two features that will need that are:

* `@JSSymbol` (#1997)
* Static fields, with their static initializers (#1902)

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0, but I suggest we do not merge this as part of this PR. Rather, we should build either `@JSSymbol` or Static fields on top of this commit to gain confidence that it addresses everything we need.

**Comment by sjrd** [comment]
Well, this causes a `SyntaxError` in PhantomJS 1.9.7 and 1.9.8, but it won't tell me where. Grrr. And it works with PhantomJS 2.1.1 (the latest).

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala [hunk 1]
```diff
@@ -2204,21 +2212,37 @@ private[emitter] class JSDesugaring(internalOptions: InternalOptions) {
...
+        if (enclosingClassName.exists(_ == className)) {
+          encodeClassVar(className)
+        } else {
+          /* In ECMAScript51Global mode, the accessor in under `ScalaJS.a`
+           * rather than `ScalaJS.c` to avoid clashes.
+           */
+          val accessor = outputMode match {
+            case OutputMode.ECMAScript51Global => envField("a", className)
+            case _                             => encodeClassVar(className)
+          }
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: Is there a reason we need to do this (and not just always use `a`)?

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala [hunk 2]
```diff
@@ -2354,7 +2386,11 @@ private[emitter] class JSDesugaring(internalOptions: InternalOptions) {
         value match {
           case js.Function(args, body) =>
             // Make sure the function has a meaningful `name` property
-            js.FunctionDef(globalVarIdent, args, body)
+            val functionExpr = js.FunctionDef(globalVarIdent, args, body)
+            if (keepFunctionExpression)
+              js.VarDef(globalVarIdent, Some(functionExpr))
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: I do not understand why this is necessary. I understand that this prevents hoisting of the function (probably, although the function keeps its name....). What does that change? It's not like the function body is evaluated...

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala [hunk 3]
```diff
@@ -113,6 +117,39 @@ private[emitter] final class ScalaJSClassEmitter(
...
+      val createClassValueVar =
+        envFieldDef("b", className, js.Undefined(), mutable = true)
+
+      val createAccessor = {
+        val classValueVar = envField("b", className)
+
+        val assignClassValueVar = js.Block(
+            entireClassDef,
+            classValueVar := envField("c", className)
+        )
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala]: Could you move this inside the `js.If` or make it somehow clear that this is executed conditionally? It's kind of hard to read top-to-bottom otherwise.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala [hunk 4]
```diff
@@ -113,6 +117,39 @@ private[emitter] final class ScalaJSClassEmitter(
...
+        val initBlock =
+          js.If(!classValueVar, assignClassValueVar, js.Skip())
+
+        val body = js.Block(initBlock, js.Return(classValueVar))
+
+        val envFieldName = outputMode match {
+          case OutputMode.ECMAScript51Global => "a"
+          case _                             => "c"
+        }
+        envFieldDef(envFieldName, className, js.Function(Nil, body))
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala]: Same here: Is the "override `c` stuff" really necessary?

File: tools/scalajsenv.js [hunk 5]
```diff
@@ -125,6 +125,8 @@ ScalaJS.clz32 = ScalaJS.g["Math"]["clz32"] || (function(i) {
 // Other fields
 //!if outputMode == ECMAScript51Global
 ScalaJS.d = {};         // Data for types
+ScalaJS.a = {};         // Scala.js-defined JS class value accessors
+ScalaJS.b = {};         // Scala.js-defined JS class value fields
 ScalaJS.c = {};         // Scala.js constructors
```
Comment (gzm0) [inline tools/scalajsenv.js]: Just do double check: There are three of these, such that during initialization of a SJS defined class, the class itself is not yet available externally.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala [hunk 6]
```diff
@@ -203,7 +203,7 @@ private[emitter] class JSDesugaring(internalOptions: InternalOptions) {
   /** Desugars parameters and body to a JS function.
    */
   private[emitter] def desugarToFunction(
-      classEmitter: ScalaJSClassEmitter, enclosingClassName: String,
+      classEmitter: ScalaJSClassEmitter, enclosingClassName: Option[String],
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: Just to double check. `enclosingClassName` is becoming optional because:

- It makes more sense.
- To prevent static methods of SJS defined JS classes from being treated specially w.r.t. their classes.

If it is not too much of a hassle, consider moving this change into a separate commit.

Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: That's right. I'll try to extract that in a separate commit. It shouldn't be too hard.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala (hunk 1 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: Not really. It's probably better to consistently use `a` indeed.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala (hunk 2 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: It caused a `ParseError` in PhantomJS if I did that differently. I'm not sure there's something against that in the ES spec (5.1, 5.1 strict and/or 2015), but I seem to remember I saw something about that someday. The current implementation is safe and assured to always work.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala (hunk 1 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: Doing so actually significantly improves the situation, as `enclosingClassName` is not necessary anymore, so I can revert all the changes I made about `enclosingClassName`.

**Comment by sjrd** [comment]
Updated.

**Comment by sjrd** [comment]
Rebased. Solved conflict in `project/BinaryIncompatibilities.scala`.

**Comment by sjrd** [comment]
This has been integrated as part of #2705.
